File: csharp/src/Structural/Composite/TinOfTeaBags.cs
//TinOfTeaBags.java - one composite extension - the "node"

namespace Structural.Composite {

using System.Collections.Generic;

public class TinOfTeaBags : TeaBags {  
   public TinOfTeaBags(string nameIn) {
       teaBagList = new List<TeaBags>();
       this.setName(nameIn);
   }
   
   override public int countTeaBags() {
       int totalTeaBags = 0;
       foreach(TeaBags tb in teaBagList) {
           totalTeaBags += tb.countTeaBags();
       }
       return totalTeaBags;
   }
   
   override public bool add(TeaBags teaBagsToAdd) {
       teaBagsToAdd.setParent(this);
       teaBagList.Add(teaBagsToAdd);
			return true;
   }
   
   override public bool remove(TeaBags teaBagsToRemove) {
       foreach(TeaBags tb in teaBagList) {
           if (tb == teaBagsToRemove) {
               teaBagList.Remove(tb);
               return true;
           }
       }
       return false;
   }
   
}

}

File: scala/src/main/scala/Structural/Composite/TinOfTeaBags.scala
package Structural.Composite;


//TinOfTeaBags - one composite extension - the "node"

import java.util.LinkedList;
import java.util.ListIterator;

class TinOfTeaBags extends TeaBags {  
   TinOfTeaBags(String nameIn) {
       teaBagList = new LinkedList();
       this.setName(nameIn);
   }
   
   int countTeaBags() {
       int totalTeaBags = 0;
       ListIterator listIterator = this.createListIterator();
       TeaBags tempTeaBags;
       while (listIterator.hasNext()) {
           tempTeaBags = (TeaBags)listIterator.next();
           totalTeaBags += tempTeaBags.countTeaBags();
       }
       return totalTeaBags;
   }
   
   boolean add(TeaBags teaBagsToAdd) {
       teaBagsToAdd.setParent(this);
       return teaBagList.add(teaBagsToAdd);
   }
   
   boolean remove(TeaBags teaBagsToRemove) {
       ListIterator listIterator = 
           this.createListIterator();
       TeaBags tempTeaBags;
       while (listIterator.hasNext()) {
           tempTeaBags = (TeaBags)listIterator.next();
           if (tempTeaBags == teaBagsToRemove) {
               listIterator.remove();
               return true;
           }
       }
       return false;
   }
   
   ListIterator createListIterator() {
       ListIterator listIterator = teaBagList.listIterator();
       return listIterator;
   }
}